File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn Earth seconds = ageOnEarth seconds
ageOn Mercury seconds = (ageOnEarth seconds) / 0.2408467
ageOn Venus seconds = (ageOnEarth seconds) / 0.61519726
ageOn Mars seconds = (ageOnEarth seconds) / 1.8808158
ageOn Jupiter seconds = (ageOnEarth seconds) / 11.862615
ageOn Saturn seconds = (ageOnEarth seconds) / 29.447498
ageOn Uranus seconds = (ageOnEarth seconds) / 84.016846
ageOn Neptune seconds = (ageOnEarth seconds) / 164.79132

ageOnEarth :: Float -> Float
ageOnEarth seconds = seconds / 365.25 / 24 / 3600

File: typescript/space-age/space-age.ts
type PlanetRatio = {
  planet_name: string;
  ratio: number;
}

export function age(planet: string, seconds: number): number {
  const earth_year_in_seconds: number = 31_557_600;
  const planet_year_to_earth_ratio: PlanetRatio[] = [
    {planet_name: 'earth', ratio: 1},
    {planet_name: 'mercury', ratio: 0.2408467},
    {planet_name: 'venus', ratio: 0.61519726},
    {planet_name: 'mars', ratio: 1.8808158},
    {planet_name: 'jupiter', ratio: 11.862615},
    {planet_name: 'saturn', ratio: 29.447498},
    {planet_name: 'uranus', ratio: 84.016846},
    {planet_name: 'neptune', ratio: 164.79132},
  ]

  const planet_ratio: number = planet_year_to_earth_ratio.find((elem) => elem.planet_name === planet)?.ratio || 1
  const unrounded_result = seconds / (earth_year_in_seconds * planet_ratio);
  return Number(unrounded_result.toFixed(2));
}
